src/compiler/checker.ts

```diff
@@ -6258,6 +6258,13 @@ namespace ts {
                     targetType = typeToString(target, /*enclosingDeclaration*/ undefined, TypeFormatFlags.UseFullyQualifiedType);
                 }
 
+                // check if trying to relate primitives to the boxed/apparent backing types.
+                if ((sourceType === "Number" && targetType === "number") ||

```

Comment:
Also, I don't think you necessarily want to do this here. I think you want to have a separate function for this logic right before this line in `isRelatedTo`:

```ts
if (reportErrors) {
    reportRelationError(headMessage, source, target);
}
```

So before that, I'd have a check like

```ts
if (source.flags & TypeFlags.ObjectType && target.flags & TypeFlags.Primitive) {
    elaborateErrorForPrimitivesAndObjects(source, target);
}
```